python/ray/tests/test_unavailable_actors.py

```diff
@@ -0,0 +1,184 @@
+
+        # Kill the actor process. The actor will restart so we get a temporal
+        # unavailable.
+        pid = ray.get(a.getpid.remote())
+        os.kill(pid, signal.SIGKILL)
+        with pytest.raises(ActorUnavailableError):
+            print(ray.get(a.slow_increment.remote(2, 0.1)))
+
+        # Actor restarting for 5s. In this period, we get a RESTARTING issue.
+        with pytest.raises(
+            ActorUnavailableError, match="The actor is in RESTARTING state"
+        ):
+            print(ray.get(a.slow_increment.remote(2, 0.1)))
+        time.sleep(6)
+
+        # After actor started, next calls are OK. However the previous actor instance's
+        # state is lost.
+        total = 0
+        for i in range(10):
+            total += i
+            assert ray.get(a.slow_increment.remote(i, 0.1)) == total
+
+        # Kill the actor again. This time it's not gonna restart so RayActorError.
+        pid = ray.get(a.getpid.remote())
+        os.kill(pid, signal.SIGKILL)
+        with pytest.raises(RayActorError):
+            print(ray.get(a.slow_increment.remote(1, 0.1)))
+
+    call_from(body, caller)
+
+
+@pytest.mark.parametrize(
+    "caller",
+    ["actor", "task", "driver"],
+)
+def test_actor_unavailable_norestart(ray_start_regular, caller):
+    def body():
+        a = Counter.remote()
+        assert ray.get(a.sum.remote(1, 2)) == 3
+        pid = ray.get(a.getpid.remote())
+
+        # Kill the actor process. The actor died permanently so RayActorError.
+        os.kill(pid, signal.SIGKILL)
+        with pytest.raises(RayActorError):
+            print(ray.get(a.sum.remote(1, 2)))
+
+    call_from(body, caller)
+
+
+@pytest.mark.skip(reason="https://github.com/ray-project/ray/issues/44357")
+@pytest.mark.parametrize(
+    "caller",
+    ["actor", "task", "driver"],
+)
+def test_generators_early_stop_unavailable(ray_start_regular, caller):
+    """
+    For a streaming generator, if a connection break happens, *some* elements may still
+    yield because they are received prior to the break. Then, two things can happen:
+    - an early StopIteration, if `next(gen)` is called.
+    - a ActorUnavailableError, if `ray.get(obj_ref)` is called.
+    """
+
+    def body():
+        a = Counter.remote()
+        pid = ray.get(a.getpid.remote())
+        total = 2000000
+        break_at = 5
+        gen = a.gen_iota.remote(total)
+        obj_refs = []
+        i = 0
+        for obj_ref in gen:
+            obj_refs.append(obj_ref)
+            if i == break_at:
+                print(f"breaking conns at {i}")
+                close_common_connections(pid)
+            i += 1
+        # StopIteration happened before `total` elements reached.
+        # On my laptop, 11120 elements are collected.
+        print(f"collected {len(obj_refs)} elements")
+        assert len(obj_refs) < total
+        with pytest.raises(ActorUnavailableError):
+            ray.get(obj_refs)
```

Comment:
Would all the object refs raise an `ActorUnavailableError` here or would the first 5 refs succeed?

Reply:
We break the connection at the 5th `ray.get` so the first 5 values and some more values (I guess up to the queued window, ~11k on my laptop) are fine and would return values.

Reply:
Got it, that makes sense.